### ISSUE TITLE
Directional Contrastive Loss calculation

### ISSUE THREAD

**Issue by yaping222** [Jia-Research-Lab/Context-Aware-Consistency#14]
Hi X-Lai, 

Thank you very much for sharing your well-organized code and I'm trying to use your Directional Contrastive Loss in my work. I have a question on the calculation of Directional Contrastive Loss. 
Equation 1 in your conference paper has two items in the denominator, while the calculation in your code is as following:
logits1 = torch.exp(pos1 - neg_max1).squeeze(-1) / (logits1_down + eps) # in model.py Line 231.
Should it be 'torch.exp(pos1 - neg_max1).squeeze(-1) / (torch.exp(pos1 - eg_max1).squeeze(-1)+logits1_down + eps)'? Or 'torch.exp(pos1 - neg_max1).squeeze(-1)' is already added up to logits1_down.

Ps: Is 'neg_max1' for normalization?

I'm looking forward to your reply!

**Comment by X-Lai** [comment]
Thanks for your interest in our work.

Yes, you're right. 'torch.exp(pos1 - neg_max1).squeeze(-1)' is already added up to logits1_down. Because in calculating the logits1_down, the positive sample will always be incorporated implicitly.

The neg_max1 is to avoid NaN in calculating the softmax function. As you know, the softmax is calculated as `exp(x_i) / (exp(x_0)+exp(x_1)+...+exp(x_n))`, however, practically, some of `exp(x_j)` may get the value of INF, causing the NaN problem. To avoid this situation, we calculate the softmax in an equivalent form as  `exp(x_i-neg_max) / (exp(x_0-neg_max)+exp(x_1-neg_max)+...+exp(x_n-neg_max))`, where the `neg_max` is the maximum of all `x_j`. In this way, any of `exp(x_j-neg_max)` won't get the value of INF. Meanwhile, the equation won't change the result of softmax function.

**Comment by yaping222** [comment]
Hi X-Lai,

Thank you very much for your explanation. There are two points still confuse me. 
1. In the paper, you mention 'we regard two features at the same location of φo1 and φo2 as a positive pair'. In 'model.py Line 171', pixels are randomly selected from the whole image, positive samples may be skipped.
2. 'mask1_idx' defined in 'model.py Line 204 and 211' suggests that only pixels with different pseudo labels are taken as a pair. This seems only the negative pairs are considered but not positive pairs, because positive pairs are supposed to have the same pseudo label.

I guess the influence of positive pairs in the denominator is quite limited when using a large amount of negative pairs.

I'm looking forward to your reply.

**Comment by X-Lai** [comment]
Yes, you're right. But in this line [link](https://github.com/dvlab-research/Context-Aware-Consistency/blob/master/models/model.py#L213), you can see we actually incorporate the positive pair manually.

**Comment by yaping222** [comment]
Hi X-Lai, I see! I missed that line before. Thank you very much!
